sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/RocksDBSuite.scala

```diff
@@ -2201,6 +2246,97 @@ class RocksDBSuite extends AlsoTestWithChangelogCheckpointingEnabled with Shared
     }
   }
 
+  testWithChangelogCheckpointingEnabled("load the same version of pending snapshot uploading") {
+    // The test was accidentally fixed by SPARK-48931 (https://github.com/apache/spark/pull/47393)
+    val remoteDir = Utils.createTempDir().toString
+    val conf = dbConf.copy(minDeltasForSnapshot = 2, compactOnCommit = false)
+    new File(remoteDir).delete() // to make sure that the directory gets created
+    withDB(remoteDir, conf = conf) { db =>
+      db.load(0)
+      db.put("foo", "bar")
+      db.commit()
+
+      db.load(1)
+      db.put("foo", "bar")
+      db.commit()
+      db.doMaintenance()
+
+      db.load(1)
+      db.put("foo", "bar")
+      db.commit()
+
+      db.load(2)
+      db.put("foo", "bar")
+      db.commit()
+
+      db.load(2)
+      db.put("foo", "bar")
+      db.commit()
+
+      db.doMaintenance()
+
+      db.load(2)
+      db.put("foo", "bar")
+      db.commit()
+    }
+  }
+
+  for (random_seed <- 1 to 16) {
+    testWithChangelogCheckpointingEnabled(s"randomized snapshotting $random_seed") {
+      // The unit test simulates the case where batches can be reloaded and maintenance tasks
+      // can be delayed. After each batch, we randomly decide whether we would move onto the
+      // next batch, and whetehr maintenance task is executed.
+      // The test was accidentally fixed by SPARK-48931 (https://github.com/apache/spark/pull/47393)
+      val remoteDir = Utils.createTempDir().toString
+      val conf = dbConf.copy(minDeltasForSnapshot = 3, compactOnCommit = false)
+      new File(remoteDir).delete() // to make sure that the directory gets created
+      withDB(remoteDir, conf = conf) { db =>
+        val random = new Random(random_seed)
+        var curVer: Int = 0
+        for (i <- 1 to 100) {
+          db.load(curVer)
+          db.put("foo", "bar")
+          db.commit()
+          // For a one in five chance, maintenance task is executed. The chance is created to
+          // simulate the case where snapshot isn't immediately uploaded, and even delayed
+          // so that the next snapshot is ready. We create a snapshot in every 3 batches, so
+          // with 1/5 chance, it is more likely to create longer maintenance delay.
+          if (random.nextInt(5) == 0) {
+            db.doMaintenance()
+          }
+          // For half the chance, we move to the next version, and half the chance we keep the
+          // same version
+          if (random.nextInt(2) == 0) {
+            curVer = curVer + 1
+          }
+        }
+      }
+    }
+  }
+
+  testWithChangelogCheckpointingEnabled(s"simulate ForEachBatch") {
```

Comment:
Maybe ditto, but this is based on deterministic random seed. Do we confirm that this test failed before SPARK-48931? I'm OK to add this separately if the seed number is known to be problematic for old Spark version, otherwise it might not worth to add this separately.

Reply:
I tested it without that change and it worked. So that change didn't fix the test. I removed that comment. But I still think it is good to have this test.

Reply:
I meant, from my understanding, this is mostly the same with @siying added with random seed, except here we pick the specific seed. If we have to do this separately, we should clarify why seed = 66 should be picked up consistently. Did we see any test failure from seed = 66?

Reply:
What is special for this test compared to the test for `randomized snapshotting $randomSeed ifTestSkipBatch $ifTestSkipBatch`? If we see test failure from this test then I agree we should probably retain the set of random seeds to cover regression. Otherwise it's just a generalization of the fuzz test and I'm not sure we'd need to have it separately.

Reply:
removed. I tested it without that change and it worked.